### DISCUSSION TITLE
Optional input with default from context

### DISCUSSION THREAD

**Discussion by zirkelc** [trpc/trpc#4361] [Q&A]
I have a field `tenantId` which is necessary for all request. However, if the React client makes a request, the JWT token that is being sent with the request contains the `tenantId` already and in this case the input field is optional. On the other, other clients (backend services) need to pass the `tenantId` explicitly as input.

Is it possible to populate the input field `tenantId`  from the context if it was provided as input?

```ts
const baseProcedure = publicProcedure
  .input(
    z.object({
      tenantId: z.string().optional(),
    }),
  )
  .use((opts) => {
    // override tenantId with context value if it wasn't provided
    opts.input.tenantId = opts.input.tenantId || opts.ctx.tenantId;

    if (!opts.input.tenantId)
      throw new TRPCError({ code: 'UNAUTHORIZED', message: 'No tenant was provided for this request' });

    // how to provide input tenantId to the following procedures?
    return opts.next(opts);
  });
```

The following procedures cannot infer that `tenantId` is always defined, because I cannot pass the modified inputs to the `next()` function.

**Comment by Nick-Lucas** [comment]
You probably want Context Extension in this case

https://trpc.io/docs/server/middlewares#context-extension 

The middleware looks like what you want, making it optional on both input/auth and validate at least one of them is set at runtime

**Reply by zirkelc** [reply]
Is there a way to assert the type, so that the following procedures are aware that the field was already checked in the middleware and it is therefore not undefined?

**Reply by zirkelc** [reply]
Here is a sandbox with an example: https://stackblitz.com/edit/trpc-trpc-ir2lkq?file=server/trpc.ts

I would like to have all routes after the middleware be certain that `tenantId` is defined.
